### DISCUSSION TITLE
How does Temperature fallback with beam search work?

### DISCUSSION THREAD

**Discussion by tom-huntington** [openai/whisper#549] [Q&A] [answered]
https://github.com/openai/whisper/blob/02aa851a4910201f0db56960064d7e121a01002c/whisper/transcribe.py#L102-L128

If we need_fallback, then we try again with the next temperature value
```
temperature: Union[float, Tuple[float, ...]] = (0.0, 0.2, 0.4, 0.6, 0.8, 1.0),
```
But the temprature values only seems to affect the `GreadyDecoder`, not the `BeamSearchDecoder`.

Does the fallback just do exactly the same thing again when decoding with beam search?

**Comment by tom-huntington** [comment]
> --beam_size BEAM_SIZE
                        number of beams in beam search, only applicable when temperature is zero (default: 5)

So the fallback must always go to greedy decoding.

**Comment by jongwook** [comment, marked as answer]
Both beam search and greedy decoding are deterministic algorithms and make sense only with temperature 0. With nonzero temperature, the implementation becomes nondeterministic and uses the `best_of` parameter, which defaults to 5 in the CLI, where it makes `best_of` independent samples and select the one with the highest log probability.

**Reply by tom-huntington** [reply]
Thanks for clearing this up! The code isn't written to make this clear

https://github.com/openai/whisper/blob/eff383b27b783e280c089475852ba83f20f64998/whisper/decoding.py#L477-L482

**Reply by jongwook** [reply]
Good point! I should consider renaming it to something like `TemperatureSampler`..

**Reply by tom-huntington** [reply]
It's doing both the greedy decoding and the indeterministic temperature decoding

**Reply by HAJEKEL** [reply]
@jongwook when you say "Both beam search and greedy decoding are deterministic algorithms and make sense only with temperature 0." I believe you are trying to demonstrate that the temperature does not affect the outcome of the algorithms right? Because for the example of beam search, you pick the n most likely logits, and these specific n logits will not change for a change in temperature.
